src/v/utils/log_hist.h

```diff
@@ -0,0 +1,184 @@
+
+#include <bit>
+#include <chrono>
+#include <cstdint>
+#include <vector>
+
+/*
+ * A histogram implementation
+ * The buckets upper bounds are powers of 2 minus 1.
+ * `first_bucket_upper_bound` therefore must be a power of 2.
+ * The number of values represented by each bucket increases by powers 2 as
+ * well.
+ *
+ * Assume `number_of_buckets` is 4 and `first_bucket_upper_bound` is 16 the
+ * bucket value ranges are;
+ *
+ * [1, 16), [16, 32), [32, 64), [64, 128)
+ *
+ * And if 1, 16, 32, and 33 are recorded the buckets will have the following
+ * counts;
+ *
+ * [1, 16)   = 1
+ * [16, 32)  = 1
+ * [32, 64)  = 2
+ * [64, 128) = 0
+ */
+template<
+  typename duration_t,
+  int number_of_buckets,
+  uint64_t first_bucket_upper_bound>
+class log_hist {
+    static_assert(
+      first_bucket_upper_bound >= 1
+        && (first_bucket_upper_bound & (first_bucket_upper_bound - 1)) == 0,
+      "first bucket bound must be power of 2");
+
+public:
+    static constexpr int first_bucket_clz = std::countl_zero(
+      first_bucket_upper_bound - 1);
+    static constexpr int first_bucket_exp = 64 - first_bucket_clz;
+
+    using clock_type = std::chrono::high_resolution_clock;
+
+    /// \brief move-only type to tracking durations
+    /// if hdr_hist ptr goes out of scope, it will detach itself
+    /// and the recording will simply be ignored.
+    class measurement : public boost::intrusive::list_base_hook<> {
+    public:
+        explicit measurement(log_hist& h)
+          : _h(std::ref(h))
+          , _begin_t(log_hist::clock_type::now()) {
+            _h.get()._probes.push_back(*this);
+        }
+        measurement(const measurement&) = delete;
+        measurement& operator=(const measurement&) = delete;
+        measurement(measurement&& o) noexcept
+          : _detached(o._detached)
+          , _trace(o._trace)
+          , _h(o._h)
+          , _begin_t(o._begin_t) {
+            o.detach_hdr_hist();
+        }
+        measurement& operator=(measurement&& o) noexcept {
+            if (this != &o) {
+                this->~measurement();
+                new (this) measurement(std::move(o));
+            }
+            return *this;
+        }
+        ~measurement() noexcept {
+            if (!_detached) {
+                _h.get()._probes.erase(_h.get()._probes.iterator_to(*this));
+                // !detached && trace
+                // do not move outside of this nested if
+                if (_trace) {
+                    _h.get().record(compute_duration());
+                }
+            }
+        }
+
+        void set_trace(bool b) { _trace = b; }
+
+    private:
+        friend log_hist;
+
+        int64_t compute_duration() const {
+            return std::chrono::duration_cast<duration_t>(
+                     log_hist::clock_type::now() - _begin_t)
+              .count();
+        }
+
+        void detach_hdr_hist() { _detached = true; }
+
+        bool _detached = false;
+        bool _trace = true;
```

Comment:
I don't fully understand the purpose of `_trace`. We don't set it to `false` anywhere?

Reply:
The code for the measurement class code was largely copied from `hdr_hist` so I can't speak for the intentions of the orginal author. However, I believe the idea is that we can abandon a latency measure at any point. A good example of when we might want to do that is in our fetch handler so we don't record the latencies of any fetches without returned data. 

This being said I'll be doing a major refactor of this class soon. In which I'll probably be removing trace as it's not used like you've said. I'll also probably remove the intrusive list for a much lighter weight solution.

Reply:
I think it's also (primarily?) for shutting down servers. In that case the underlying hdr_hist assoicated with the problem will probably be destroyed at some point but you might have latency measurement objects live in random tasks throughput the system: this "detach" approach means they won't crash. Kind of a good idea since the alternative is have tricky rules about lifetime, like the kafka server needs to outlive any task which might reference these objects: perhaps easy if we already drain requests in progress but once you start passing them into other subsystems it becomes harder to guarantee.

Seems like lw_shared_ptr could be used to a similar thing with slightly different semantics.

Reply:
I just switched it over to a `lw_shard_ptr` lol. Nice call